lua/lspconfig/util.lua

```diff
@@ -347,6 +361,10 @@ function M.root_pattern(...)
     end
   end
   return function(startpath)
+    local server_name = get_config_by_ft()
+    if server_name and not M.server_root_pattern[server_name] then
+      M.server_root_pattern[server_name] = patterns
+    end
```

Comment:
Not all server configuration's `root_dir` property makes use of this `root_pattern()` utility function. I think it'll be pretty complicated to try to capture the `root_dir` predicate logic this way. A root dir function also doesn't necessarily have to deal directly with file paths (although this is most often the case) - how would these "patterns" be registered? For example (very contrived example):

```lua
lspconfig.some_server.setup {
  root_dir = function (fname)
     return sync_spawn("server-tool find-project-root", {
         cwd = vim.fn.fnamemodify(fname, "%:h")
     })
  end
}
```

I was thinking we could still simply make use of the `.docs.default_config.root_dir` value, if available. This is also what's displayed in `server_configurations.md` document that many people refer to. This is basically what was done before, but this time we can skip displaying it if we see that the user has provided a custom `root_dir`. I opened a draft PR for this alternative approach: https://github.com/neovim/nvim-lspconfig/pull/2288.

Reply:
nope. root_dir is wrong design it It doesn't meet the specification in my opinion. I am trying remake it in core. then there need some change in lspconfig. but it's break change.

Reply:
nope. root_dir is wrong design it It doesn't meet the specification in my opinion. I am trying remake it in core. then there need some change in lspconfig. but it's break change.

Reply:
back to pr. for this custom situation can we render a message like `Searched for your custom root pattern`. if user not use the root_patterns